value/src/main/java/com/google/auto/value/AutoValueExtension.java

```diff
@@ -0,0 +1,33 @@
+package com.google.auto.value;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Created by rharter on 5/1/15.
+ */
+public interface AutoValueExtension {
+
+  public interface Context {
+    ProcessingEnvironment processingEnvironment();
+    String packageName();
+    TypeElement autoValueClass();
+    Map<String, ExecutableElement> properties();
+  }
+
+  public interface GeneratedClass {
+    String className();
+    String source();
+    Collection<String> additionalImports();
+    Collection<ExecutableElement> consumedProperties();
+    Collection<String> additionalInterfaces();
+    Collection<String> additionalCode();
```

Comment:
Ryan, I just wanted to chime in and say how much Eamonn and I appreciate
the work you're doing and the forward momentum you're providing for this
important feature!



On Mon, May 18, 2015 at 12:43 PM, Ryan Harter <notifications@github.com>
wrote:

> In value/src/main/java/com/google/auto/value/AutoValueExtension.java
> <https://github.com/google/auto/pull/237#discussion_r30540505>:
>
> > +public interface AutoValueExtension {
> > +
> > +  public interface Context {
> > +    ProcessingEnvironment processingEnvironment();
> > +    String packageName();
> > +    TypeElement autoValueClass();
> > +    Map<String, ExecutableElement> properties();
> > +  }
> > +
> > +  public interface GeneratedClass {
> > +    String className();
> > +    String source();
> > +    Collection<String> additionalImports();
> > +    Collection<ExecutableElement> consumedProperties();
> > +    Collection<String> additionalInterfaces();
> > +    Collection<String> additionalCode();
>
> Yeah, I don't like the JavaPoet approach, it also adds an extra layer of
> requirements around implementation details on the extension developer.
>
> I think this discussion is good, and it seems to point to your original
> inheritance model, with a mechanism to require an extension be at the end
> of the chain, and a limit that only one such extension can be present at a
> time.
>
> Do you think it makes sense for me to note these finding in the issue,
> cancel this PR, and alter the implementation to remove the additional code
> and force extensions to create subclasses of the AutoValue generated class,
> then submit a new PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/auto/pull/237/files#r30540505>.
>



-- 
Kevin Bourrillion | Java Librarian | Google, Inc. | kevinb@google.com